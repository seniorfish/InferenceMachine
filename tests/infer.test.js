import { expect, test } from "vitest";
import { Solver, makeConversionList, formulaToElements } from "../src/infer.js";

test("makeConversionList", () => {
    expect([...new Set(makeConversionList())].sort()).toMatchSnapshot();
});
test("formulaToElements", () => {
    let o = {
        "Cu(OH)2": ["Cu", "O", "H"],
        H2O: ["H", "O"],
        "Ca(HCO3)2": ["Ca", "H", "C", "O"],
    };
    for (let [k, v] of Object.entries(o)) {
        expect([...formulaToElements(k)].sort().join()).toEqual(
            v.sort().join()
        );
    }
});

const unique = [
    `a+b=d+e
b+c=e+f
d>f<g<f
e:NaCl
f:[CO2,CO]
g:[CO2,CO]
d:CaCO3
`,
    `g-d>a<b>c<f-i<h-g
d-b-f>h<d
b>e>h
d>e>f
b:[Cu,Fe2O3]
[ghef](1)
[adbci](O)
`,
    `a-e>d-f>c<b=a
e-b
e>c
c=d
e:HCl
B:NaOH
f:Fe2O3
`,
    `d=h=g-a>b-c>d-e-f>a
[abcdefgh]{HCONaSCu}
[agh](O)
E:[H2SO4,HNO3,HCl]`,
];
const mult = [
    `b+d=e
a+e=f
a+b=c
c+d=f
[def](1)
c:[CuSO4,Ca(OH)2]
b:H2SO4
a:[Fe,C,CuO,MnO2,Fe3O4]
`,
    `c<a>b
b=c>e=d>b
c-d-f-e
[abcdef](2-3)
b:H2O
f:HCl
d:Na2CO3
`,
    `f-e-a-d-c=b<a>g=f
g:H2
a:HCl
[bcdef](O)
`,
    `a+b+c=d
[abc]{CaHCO}
`,
    `c>a
c-a`,
`c+e=f
e+d=g
f>h-e
g>i-e
c>a>b>d
e:H2SO4
c:Fe2O3
b:Cu
[g][Cu]
[cd](O)
[ab](1)
[f][Fe]
[h][HICu]
`
];

test("Solve", () => {
    for (let c of unique) {
        let a = new Solver();
        c.split("\n").map((v) => a.input(v));
        let ans = a.solve();
        expect(ans.length).toBe(1);
    }
    for (let c of mult) {
        let a = new Solver();
        c.split("\n").map((v) => a.input(v));
        let ans = a.solve();
        expect(
            ans.map((v) =>
                    [...v.entries()]
                        .map((x) => x.join(":"))
                        .sort()
                        .join(",")
                )
                .sort()
                .join(";")
        ).toMatchSnapshot();
    }
    // expect((new Solver()).solve().length).toBe(1)
});
