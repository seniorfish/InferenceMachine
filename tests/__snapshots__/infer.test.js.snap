// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Solve 1`] = `"A:CuO,B:H2SO4,C:CuSO4,D:Al,E:H2,F:Cu;A:CuO,B:H2SO4,C:CuSO4,D:Fe,E:H2,F:Cu;A:CuO,B:H2SO4,C:CuSO4,D:Mg,E:H2,F:Cu;A:CuO,B:H2SO4,C:CuSO4,D:Zn,E:H2,F:Cu"`;

exports[`Solve 2`] = `"A:Ba(OH)2,B:H2O,C:Ca(OH)2,D:Na2CO3,E:NaOH,F:HCl;A:CaO,B:H2O,C:Ca(OH)2,D:Na2CO3,E:NaOH,F:HCl"`;

exports[`Solve 3`] = `"A:HCl,B:CO2,C:CO,D:CuO,E:CaO,F:H2O,G:H2;A:HCl,B:CO2,C:CO,D:Fe2O3,E:CaO,F:H2O,G:H2"`;

exports[`Solve 4`] = `"A:CO2,B:CaCO3,C:H2O,D:Ca(HCO3)2;A:CO2,B:H2O,C:CaCO3,D:Ca(HCO3)2;A:CaCO3,B:CO2,C:H2O,D:Ca(HCO3)2;A:CaCO3,B:H2O,C:CO2,D:Ca(HCO3)2;A:H2O,B:CO2,C:CaCO3,D:Ca(HCO3)2;A:H2O,B:CaCO3,C:CO2,D:Ca(HCO3)2"`;

exports[`Solve 5`] = `"A:CO,C:O2;A:CO2,C:C;A:CO2,C:CaCO3;A:CO2,C:Na2CO3;A:CaCO3,C:CO2;A:H2O,C:CO2;A:H2O,C:CaCO3;A:H2O,C:CaO;A:H2O,C:Na2CO3;A:H2O,C:O2;A:Na2CO3,C:CO2;A:O2,C:H2O;A:SO2,C:O2"`;

exports[`Solve 6`] = `"A:Fe,B:Cu,C:Fe2O3,D:CuO,E:H2SO4,F:Fe2(SO4)3,G:CuSO4,H:Fe(OH)3,I:Cu(OH)2;A:Fe,B:Cu,C:Fe2O3,D:CuO,E:H2SO4,F:Fe2(SO4)3,G:CuSO4,H:Fe(OH)3,I:CuCO3"`;

exports[`makeConversionList 1`] = `
[
  "(NH4)2SO4+NaOH=NaSO4+NH3+H2O",
  "Ag2SO4+(NH4)2CO3=Ag2CO3+(NH4)2SO4",
  "Ag2SO4+K2CO3=Ag2CO3+K2SO4",
  "Ag2SO4+MgCO3=Ag2CO3+MgSO4",
  "Ag2SO4+Na2CO3=Ag2CO3+Na2SO4",
  "AgNO3+(NH4)2CO3=Ag2CO3+NH4NO3",
  "AgNO3+AlCl3=AgCl+Al(NO3)3",
  "AgNO3+BaCl2=AgCl+Ba(NO3)2",
  "AgNO3+CaCl2=AgCl+Ca(NO3)2",
  "AgNO3+CuCl2=AgCl+Cu(NO3)2",
  "AgNO3+FeCl3=AgCl+Fe(NO3)3",
  "AgNO3+HCl=AgCl+HNO3",
  "AgNO3+K2CO3=Ag2CO3+KNO3",
  "AgNO3+KCl=AgCl+KNO3",
  "AgNO3+MgCO3=Ag2CO3+Mg(NO3)2",
  "AgNO3+MgCl2=AgCl+Mg(NO3)2",
  "AgNO3+NH4Cl=AgCl+NH4NO3",
  "AgNO3+Na2CO3=Ag2CO3+NaNO3",
  "AgNO3+NaCl=AgCl+NaNO3",
  "Al(OH)3+H2SO4=Al2(SO4)3+H2O",
  "Al(OH)3+HCl=AlCl3+H2O",
  "Al(OH)3+HNO3=Al(NO3)3+H2O",
  "Al+AgNO3=Al(NO3)3+Ag",
  "Al+Cu(NO3)2=Al(NO3)3+Cu",
  "Al+CuSO4=Al2(SO4)3+Cu",
  "Al+Fe(NO3)2=Al(NO3)3+Fe",
  "Al+FeCl2=AlCl3+Fe",
  "Al+FeSO4=Al2(SO4)3+Fe",
  "Al+H2SO4=Al2(SO4)3+H2",
  "Al+HCl=AlCl3+H2",
  "Al+O2=Al2O3",
  "Al+Zn(NO3)2=Al(NO3)3+Zn",
  "Al+ZnCl2=AlCl3+Zn",
  "Al+ZnSO4=Al2(SO4)3+Zn",
  "AlCl3+Ag2SO4=Al2(SO4)3+AgCl",
  "Ba(NO3)2+(NH4)2CO3=BaCO3+NH4NO3",
  "Ba(NO3)2+(NH4)2SO4=BaSO4+NH4NO3",
  "Ba(NO3)2+Ag2SO4=BaSO4+AgNO3",
  "Ba(NO3)2+Al2(SO4)3=BaSO4+Al(NO3)3",
  "Ba(NO3)2+CaSO4=BaSO4+Ca(NO3)2",
  "Ba(NO3)2+CuSO4=BaSO4+Cu(NO3)2",
  "Ba(NO3)2+Fe2(SO4)3=BaSO4+Fe(NO3)3",
  "Ba(NO3)2+H2SO4=BaSO4+HNO3",
  "Ba(NO3)2+K2CO3=BaCO3+KNO3",
  "Ba(NO3)2+K2SO4=BaSO4+KNO3",
  "Ba(NO3)2+MgCO3=BaCO3+Mg(NO3)2",
  "Ba(NO3)2+MgSO4=BaSO4+Mg(NO3)2",
  "Ba(NO3)2+Na2CO3=BaCO3+NaNO3",
  "Ba(NO3)2+Na2SO4=BaSO4+NaNO3",
  "Ba(OH)2+(NH4)2CO3=BaCO3+NH3+H2O",
  "Ba(OH)2+(NH4)2SO4=BaSO4+NH3+H2O",
  "Ba(OH)2+Al(NO3)3=Ba(NO3)2+Al(OH)3",
  "Ba(OH)2+Al2(SO4)3=BaSO4+Al(OH)3",
  "Ba(OH)2+AlCl3=BaCl2+Al(OH)3",
  "Ba(OH)2+CaSO4=BaSO4+Ca(OH)2",
  "Ba(OH)2+Cu(NO3)2=Ba(NO3)2+Cu(OH)2",
  "Ba(OH)2+CuCl2=BaCl2+Cu(OH)2",
  "Ba(OH)2+CuSO4=BaSO4+Cu(OH)2",
  "Ba(OH)2+Fe(NO3)3=Ba(NO3)2+Fe(OH)3",
  "Ba(OH)2+Fe2(SO4)3=BaSO4+Fe(OH)3",
  "Ba(OH)2+FeCl3=BaCl2+Fe(OH)3",
  "Ba(OH)2+H2SO4=BaSO4+H2O",
  "Ba(OH)2+HCl=BaCl2+H2O",
  "Ba(OH)2+HNO3=Ba(NO3)2+H2O",
  "Ba(OH)2+K2CO3=BaCO3+KOH",
  "Ba(OH)2+K2SO4=BaSO4+KOH",
  "Ba(OH)2+Mg(NO3)2=Ba(NO3)2+Mg(OH)2",
  "Ba(OH)2+MgCO3=BaCO3+Mg(OH)2",
  "Ba(OH)2+MgCl2=BaCl2+Mg(OH)2",
  "Ba(OH)2+MgSO4=BaSO4+Mg(OH)2",
  "Ba(OH)2+NH4Cl=BaCl2+NH3+H2O",
  "Ba(OH)2+NH4NO3=Ba(NO3)2+NH3+H2O",
  "Ba(OH)2+Na2CO3=BaCO3+NaOH",
  "Ba(OH)2+Na2SO4=BaSO4+NaOH",
  "Ba(OH)2+SO3=BaSO4+H2O",
  "BaCl2+(NH4)2CO3=BaCO3+NH4Cl",
  "BaCl2+(NH4)2SO4=BaSO4+NH4Cl",
  "BaCl2+Ag2SO4=BaSO4+AgCl",
  "BaCl2+Al2(SO4)3=BaSO4+AlCl3",
  "BaCl2+CaSO4=BaSO4+CaCl2",
  "BaCl2+CuSO4=BaSO4+CuCl2",
  "BaCl2+Fe2(SO4)3=BaSO4+FeCl3",
  "BaCl2+H2SO4=BaSO4+HCl",
  "BaCl2+K2CO3=BaCO3+KCl",
  "BaCl2+K2SO4=BaSO4+KCl",
  "BaCl2+MgCO3=BaCO3+MgCl2",
  "BaCl2+MgSO4=BaSO4+MgCl2",
  "BaCl2+Na2CO3=BaCO3+NaCl",
  "BaCl2+Na2SO4=BaSO4+NaCl",
  "C+CuO=Cu+CO2",
  "C+Fe2O3=Fe+CO2",
  "C+Fe3O4=Fe+CO2",
  "C+O2=CO",
  "C+O2=CO2",
  "C2H5OH+O2=CO2+H2O",
  "C6H12O6+O2=CO2+H2O",
  "CH4+O2=CO2+H2O",
  "CO+CuO=Cu+CO2",
  "CO+Fe2O3=Fe+CO2",
  "CO+Fe3O4=Fe+CO2",
  "CO+O2=CO2",
  "CO2+C=CO",
  "CO2+H2O=C6H12O6+O2",
  "Ca(HCO3)2=CaCO3+CO2+H2O",
  "Ca(NO3)2+(NH4)2CO3=CaCO3+NH4NO3",
  "Ca(NO3)2+K2CO3=CaCO3+KNO3",
  "Ca(NO3)2+MgCO3=CaCO3+Mg(NO3)2",
  "Ca(NO3)2+Na2CO3=CaCO3+NaNO3",
  "Ca(OH)2+(NH4)2CO3=CaCO3+NH3+H2O",
  "Ca(OH)2+(NH4)2SO4=CaSO4+NH3+H2O",
  "Ca(OH)2+Al(NO3)3=Ca(NO3)2+Al(OH)3",
  "Ca(OH)2+Al2(SO4)3=CaSO4+Al(OH)3",
  "Ca(OH)2+AlCl3=CaCl2+Al(OH)3",
  "Ca(OH)2+CO2=CaCO3+H2O",
  "Ca(OH)2+Cu(NO3)2=Ca(NO3)2+Cu(OH)2",
  "Ca(OH)2+CuCl2=CaCl2+Cu(OH)2",
  "Ca(OH)2+CuSO4=CaSO4+Cu(OH)2",
  "Ca(OH)2+CuSO4=Cu(OH)2+CaSO4",
  "Ca(OH)2+Fe(NO3)3=Ca(NO3)2+Fe(OH)3",
  "Ca(OH)2+Fe2(SO4)3=CaSO4+Fe(OH)3",
  "Ca(OH)2+FeCl3=CaCl2+Fe(OH)3",
  "Ca(OH)2+H2SO4=CaSO4+H2O",
  "Ca(OH)2+HCl=CaCl2+H2O",
  "Ca(OH)2+HNO3=Ca(NO3)2+H2O",
  "Ca(OH)2+K2CO3=CaCO3+KOH",
  "Ca(OH)2+Mg(NO3)2=Ca(NO3)2+Mg(OH)2",
  "Ca(OH)2+MgCO3=CaCO3+Mg(OH)2",
  "Ca(OH)2+MgCl2=CaCl2+Mg(OH)2",
  "Ca(OH)2+MgSO4=CaSO4+Mg(OH)2",
  "Ca(OH)2+NH4Cl=CaCl2+NH3+H2O",
  "Ca(OH)2+NH4NO3=Ca(NO3)2+NH3+H2O",
  "Ca(OH)2+Na2CO3=CaCO3+NaOH",
  "Ca(OH)2+SO3=CaSO4+H2O",
  "CaCO3+CO2+H2O=Ca(HCO3)2",
  "CaCO3+H2SO4=CaSO4+H2O+CO2",
  "CaCO3+HCl=CaCl2+H2O+CO2",
  "CaCO3=CaO+CO2",
  "CaCl2+(NH4)2CO3=CaCO3+NH4Cl",
  "CaCl2+Ag2SO4=CaSO4+AgCl",
  "CaCl2+K2CO3=CaCO3+KCl",
  "CaCl2+MgCO3=CaCO3+MgCl2",
  "CaCl2+Na2CO3=CaCO3+NaCl",
  "CaO+H2O=Ca(OH)2",
  "CaO+H2SO4=CaSO4+H2O",
  "CaO+HCl=CaCl2+H2O",
  "CaSO4+(NH4)2CO3=CaCO3+(NH4)2SO4",
  "CaSO4+K2CO3=CaCO3+K2SO4",
  "CaSO4+MgCO3=CaCO3+MgSO4",
  "CaSO4+Na2CO3=CaCO3+Na2SO4",
  "Cu(NO3)2+(NH4)2CO3=CuCO3+NH4NO3",
  "Cu(NO3)2+K2CO3=CuCO3+KNO3",
  "Cu(NO3)2+MgCO3=CuCO3+Mg(NO3)2",
  "Cu(NO3)2+Na2CO3=CuCO3+NaNO3",
  "Cu(OH)2+H2SO4=CuSO4+H2O",
  "Cu(OH)2+HCl=CuCl2+H2O",
  "Cu(OH)2+HNO3=Cu(NO3)2+H2O",
  "Cu(OH)2=CuO+H2O",
  "Cu+AgNO3=Cu(NO3)2+Ag",
  "Cu+O2=CuO",
  "CuCl2+(NH4)2CO3=CuCO3+NH4Cl",
  "CuCl2+Ag2SO4=CuSO4+AgCl",
  "CuCl2+K2CO3=CuCO3+KCl",
  "CuCl2+MgCO3=CuCO3+MgCl2",
  "CuCl2+Na2CO3=CuCO3+NaCl",
  "CuO+H2SO4=CuSO4+H2O",
  "CuO+HCl=CuCl2+H2O",
  "CuSO4+(NH4)2CO3=CuCO3+(NH4)2SO4",
  "CuSO4+Ba(OH)2=Cu(OH)2+BaSO4",
  "CuSO4+BaCl2=CuCl2+BaSO4",
  "CuSO4+K2CO3=CuCO3+K2SO4",
  "CuSO4+MgCO3=CuCO3+MgSO4",
  "CuSO4+Na2CO3=CuCO3+Na2SO4",
  "CuSO4+NaOH=Cu(OH)2+Na2SO4",
  "Fe(NO3)3+(NH4)2CO3=Fe2(CO3)3+NH4NO3",
  "Fe(NO3)3+K2CO3=Fe2(CO3)3+KNO3",
  "Fe(NO3)3+MgCO3=Fe2(CO3)3+Mg(NO3)2",
  "Fe(NO3)3+Na2CO3=Fe2(CO3)3+NaNO3",
  "Fe(OH)3+H2SO4=Fe2(SO4)3+H2O",
  "Fe(OH)3+HCl=FeCl3+H2O",
  "Fe(OH)3+HNO3=Fe(NO3)3+H2O",
  "Fe+AgNO3=Fe(NO3)2+Ag",
  "Fe+Cu(NO3)2=Fe(NO3)2+Cu",
  "Fe+CuSO4=FeSO4+Cu",
  "Fe+H2SO4=FeSO4+H2",
  "Fe+HCl=FeCl2+H2",
  "Fe+O2+H2O=Fe2O3",
  "Fe+O2=Fe3O4",
  "Fe2(SO4)3+(NH4)2CO3=Fe2(CO3)3+(NH4)2SO4",
  "Fe2(SO4)3+K2CO3=Fe2(CO3)3+K2SO4",
  "Fe2(SO4)3+MgCO3=Fe2(CO3)3+MgSO4",
  "Fe2(SO4)3+Na2CO3=Fe2(CO3)3+Na2SO4",
  "Fe2O3+H2SO4=Fe2(SO4)3+H2O",
  "Fe2O3+HCl=FeCl3+H2O",
  "FeCl3+(NH4)2CO3=Fe2(CO3)3+NH4Cl",
  "FeCl3+Ag2SO4=Fe2(SO4)3+AgCl",
  "FeCl3+K2CO3=Fe2(CO3)3+KCl",
  "FeCl3+MgCO3=Fe2(CO3)3+MgCl2",
  "FeCl3+Na2CO3=Fe2(CO3)3+NaCl",
  "H2+Cl2=HCl",
  "H2+CuO=Cu+H2O",
  "H2+Fe2O3=Fe+H2O",
  "H2+Fe3O4=Fe+H2O",
  "H2+O2=H2O",
  "H2CO3=H2O+CO2",
  "H2O+CO2=H2CO3",
  "H2O2=H2O+O2",
  "H2O=H2+O2",
  "H2SO4+(NH4)2CO3=H2O+CO2+(NH4)2SO4",
  "H2SO4+Ag2CO3=H2O+CO2+Ag2SO4",
  "H2SO4+Al2O3=H2O+Al2(SO4)3",
  "H2SO4+BaCO3=H2O+CO2+BaSO4",
  "H2SO4+BaCl2=BaSO4+HCl",
  "H2SO4+Ca(OH)2=CaSO4+H2O",
  "H2SO4+CaCO3=H2O+CO2+CaSO4",
  "H2SO4+CaO=H2O+CaSO4",
  "H2SO4+CuCO3=H2O+CO2+CuSO4",
  "H2SO4+CuO=H2O+CuSO4",
  "H2SO4+Fe2(CO3)3=H2O+CO2+Fe2(SO4)3",
  "H2SO4+Fe2O3=H2O+Fe2(SO4)3",
  "H2SO4+K2CO3=H2O+CO2+K2SO4",
  "H2SO4+MgCO3=H2O+CO2+MgSO4",
  "H2SO4+Na2CO3=H2O+CO2+Na2SO4",
  "H2SO4+NaOH=Na2SO4+H2O",
  "HCl+(NH4)2CO3=H2O+CO2+NH4Cl",
  "HCl+Ag2CO3=H2O+CO2+AgCl",
  "HCl+Ag2SO4=H2SO4+AgCl",
  "HCl+AgNO3=AgCl+HNO3",
  "HCl+Al(OH)3=AlCl3=H2O",
  "HCl+Al2O3=H2O+AlCl3",
  "HCl+BaCO3=H2O+CO2+BaCl2",
  "HCl+Ca(OH)2=CaCl2+H2O",
  "HCl+CaCO3=H2O+CO2+CaCl2",
  "HCl+CaO=H2O+CaCl2",
  "HCl+CuCO3=H2O+CO2+CuCl2",
  "HCl+CuO=H2O+CuCl2",
  "HCl+Fe2(CO3)3=H2O+CO2+FeCl3",
  "HCl+Fe2O3=H2O+FeCl3",
  "HCl+K2CO3=H2O+CO2+KCl",
  "HCl+Mg(OH)2=MgCl2+H2O",
  "HCl+MgCO3=H2O+CO2+MgCl2",
  "HCl+Na2CO3=H2O+CO2+NaCl",
  "HCl+NaOH=NaCl+H2O",
  "HNO3+(NH4)2CO3=H2O+CO2+NH4NO3",
  "HNO3+Ag2CO3=H2O+CO2+AgNO3",
  "HNO3+Al2O3=H2O+Al(NO3)3",
  "HNO3+BaCO3=H2O+CO2+Ba(NO3)2",
  "HNO3+CaCO3=H2O+CO2+Ca(NO3)2",
  "HNO3+CaO=H2O+Ca(NO3)2",
  "HNO3+CuCO3=H2O+CO2+Cu(NO3)2",
  "HNO3+CuO=H2O+Cu(NO3)2",
  "HNO3+Fe2(CO3)3=H2O+CO2+Fe(NO3)3",
  "HNO3+Fe2O3=H2O+Fe(NO3)3",
  "HNO3+K2CO3=H2O+CO2+KNO3",
  "HNO3+MgCO3=H2O+CO2+Mg(NO3)2",
  "HNO3+Na2CO3=H2O+CO2+NaNO3",
  "Hg+O2=HgO",
  "HgO=Hg+O2",
  "KCl+Ag2SO4=K2SO4+AgCl",
  "KClO3=KCl+O2",
  "KMnO4=K2MnO4+MnO2+O2",
  "KOH+(NH4)2CO3=K2CO3+NH3+H2O",
  "KOH+(NH4)2SO4=K2SO4+NH3+H2O",
  "KOH+Al(NO3)3=KNO3+Al(OH)3",
  "KOH+Al2(SO4)3=K2SO4+Al(OH)3",
  "KOH+AlCl3=KCl+Al(OH)3",
  "KOH+Cu(NO3)2=KNO3+Cu(OH)2",
  "KOH+CuCl2=KCl+Cu(OH)2",
  "KOH+CuSO4=K2SO4+Cu(OH)2",
  "KOH+Fe(NO3)3=KNO3+Fe(OH)3",
  "KOH+Fe2(SO4)3=K2SO4+Fe(OH)3",
  "KOH+FeCl3=KCl+Fe(OH)3",
  "KOH+H2SO4=K2SO4+H2O",
  "KOH+HCl=KCl+H2O",
  "KOH+HNO3=KNO3+H2O",
  "KOH+Mg(NO3)2=KNO3+Mg(OH)2",
  "KOH+MgCO3=K2CO3+Mg(OH)2",
  "KOH+MgCl2=KCl+Mg(OH)2",
  "KOH+MgSO4=K2SO4+Mg(OH)2",
  "KOH+NH4Cl=KCl+NH3+H2O",
  "KOH+NH4NO3=KNO3+NH3+H2O",
  "KOH+SO3=K2SO4+H2O",
  "Mg(OH)2+H2SO4=MgSO4+H2O",
  "Mg(OH)2+HCl=MgCl2+H2O",
  "Mg(OH)2+HNO3=Mg(NO3)2+H2O",
  "Mg+Al2(SO4)3=MgSO4+Al",
  "Mg+AlCl3=MgCl2+Al",
  "Mg+CuSO4=MgSO4+Cu",
  "Mg+FeCl2=MgCl2+Fe",
  "Mg+FeSO4=MgSO4+Fe",
  "Mg+H2SO4=MgSO4+H2",
  "Mg+HCl=MgCl2+H2",
  "Mg+O2=MgO",
  "Mg+ZnCl2=MgCl2+Zn",
  "Mg+ZnSO4=MgSO4+Zn",
  "MgCl2+Ag2SO4=MgSO4+AgCl",
  "NH4Cl+Ag2SO4=(NH4)2SO4+AgCl",
  "NH4Cl+NaOH=NaCl+NH3+H2O",
  "NH4HCO3=NH3+CO2+H2O",
  "NH4NO3+Ca(OH)2=Ca(NO3)2+NH3+H2O",
  "Na+Cl2=NaCl",
  "Na2CO3+BaCl2=NaCl+BaCO3",
  "Na2CO3+CO2+H2O=NaHCO3",
  "Na2CO3+Ca(OH)2=NaOH+CaCO3",
  "Na2CO3+CaCl2=NaCl+CaCO3",
  "Na2CO3+HCl=NaCl+H2O+CO2",
  "NaCl+Ag2SO4=Na2SO4+AgCl",
  "NaHCO3+HCl=NaCl+H2O+CO2",
  "NaHCO3=Na2CO3+CO2+H2O",
  "NaOH+(NH4)2CO3=Na2CO3+NH3+H2O",
  "NaOH+(NH4)2SO4=Na2SO4+NH3+H2O",
  "NaOH+Al(NO3)3=NaNO3+Al(OH)3",
  "NaOH+Al2(SO4)3=Na2SO4+Al(OH)3",
  "NaOH+AlCl3=NaCl+Al(OH)3",
  "NaOH+CO2=Na2CO3+H2O",
  "NaOH+Cu(NO3)2=NaNO3+Cu(OH)2",
  "NaOH+CuCl2=NaCl+Cu(OH)2",
  "NaOH+CuSO4=Cu(OH)2+Na2SO4",
  "NaOH+CuSO4=Na2SO4+Cu(OH)2",
  "NaOH+Fe(NO3)3=NaNO3+Fe(OH)3",
  "NaOH+Fe2(SO4)3=Na2SO4+Fe(OH)3",
  "NaOH+FeCl3=Fe(OH)3+NaCl",
  "NaOH+FeCl3=NaCl+Fe(OH)3",
  "NaOH+H2SO4=Na2SO4+H2O",
  "NaOH+HCl=NaCl+H2O",
  "NaOH+HNO3=NaNO3+H2O",
  "NaOH+Mg(NO3)2=NaNO3+Mg(OH)2",
  "NaOH+MgCO3=Na2CO3+Mg(OH)2",
  "NaOH+MgCl2=Mg(OH)2+NaCl",
  "NaOH+MgCl2=NaCl+Mg(OH)2",
  "NaOH+MgSO4=Na2SO4+Mg(OH)2",
  "NaOH+NH4Cl=NaCl+NH3+H2O",
  "NaOH+NH4NO3=NaNO3+NH3+H2O",
  "NaOH+SO2=Na2SO3+H2O",
  "NaOH+SO3=Na2SO4+H2O",
  "P+O2=P2O5",
  "S+O2=SO2",
  "SO2+O2=SO3",
  "Zn+AgNO3=Zn(NO3)2+Ag",
  "Zn+Cu(NO3)2=Zn(NO3)2+Cu",
  "Zn+CuSO4=ZnSO4+Cu",
  "Zn+Fe(NO3)2=Zn(NO3)2+Fe",
  "Zn+FeCl2=ZnCl2+Fe",
  "Zn+FeSO4=ZnSO4+Fe",
  "Zn+H2SO4=ZnSO4+H2",
  "Zn+HCl=ZnCl2+H2",
]
`;
